src/tabwidget.cpp

```diff
@@ -512,33 +523,52 @@ void TabWidget::preset2Horizontal()
     TerminalConfig defaultConfig;
     int ix = TabWidget::addNewTab(defaultConfig);
     TermWidgetHolder* term = reinterpret_cast<TermWidgetHolder*>(widget(ix));
+
+    // NOTE: When splitting happens, the focus changes. Therefore, we should switch to
+    // the 1st terminal only when the window is activated and the focus has really changed.
+    QObject::disconnect(mFocusConnection);
+    mFocusConnection = connect(term, &TermWidgetHolder::termFocusChanged, this, [this, term] {
+        QObject::disconnect(mFocusConnection);
+        term->directionalNavigation(NavigationDirection::Top);
+    });
+
     term->splitHorizontal(term->currentTerminal());
-    // switch to the 1st terminal
-    term->directionalNavigation(NavigationDirection::Left);
 }
 
 void TabWidget::preset2Vertical()
 {
     TerminalConfig defaultConfig;
     int ix = TabWidget::addNewTab(defaultConfig);
     TermWidgetHolder* term = reinterpret_cast<TermWidgetHolder*>(widget(ix));
+
+    // see preset2Horizontal() for an explanation
+    QObject::disconnect(mFocusConnection);
+    mFocusConnection = connect(term, &TermWidgetHolder::termFocusChanged, this, [this, term] {
+        QObject::disconnect(mFocusConnection);
+        term->directionalNavigation(NavigationDirection::Left);
+    });
+
     term->splitVertical(term->currentTerminal());
-    // switch to the 1st terminal
-    term->directionalNavigation(NavigationDirection::Left);
 }
 
 void TabWidget::preset4Terminals()
 {
     TerminalConfig defaultConfig;
     int ix = TabWidget::addNewTab(defaultConfig);
     TermWidgetHolder* term = reinterpret_cast<TermWidgetHolder*>(widget(ix));
-    term->splitVertical(term->currentTerminal());
-    term->splitHorizontal(term->currentTerminal());
-    term->directionalNavigation(NavigationDirection::Left);
 
-    term->splitHorizontal(term->currentTerminal());
-    // switch to the 1st terminal
-    term->directionalNavigation(NavigationDirection::Top);
+    // see preset2Horizontal() for an explanation
+    QObject::disconnect(mFocusConnection);
+    mFocusConnection = connect(term, &TermWidgetHolder::termFocusChanged, this, [this, term] {
+        QObject::disconnect(mFocusConnection);
+        term->splitHorizontal(term->currentTerminal());
+        term->directionalNavigation(NavigationDirection::Left);
+        term->splitHorizontal(term->currentTerminal());
+        // switch to the 1st terminal (the focus is already changed)
```

Comment:
Is this statement always true? Otherwise, the focus may be at the left-bottom subterminal if the next line (`directionalNavigation`) is run before the terminal is actually split. Not a big deal, though.

Reply:
> Is this statement always true?

If you mean "the focus is already changed", yes, it's always true at this stage. Our problem was that we didn't wait for the window activation. By waiting for `termFocusChanged`, we guarantee that the focus is changed correctly *afterward*.

Reply:
> If you mean "the focus is already changed"

Yep

> By waiting for `termFocusChanged`, we guarantee that the focus is changed correctly _afterward_.

I think only the focus change signal from `splitVertical` is waited. The focus change from the second `splitHorizontal` call in the inner lambda function is not?

Reply:
Waiting for the first focus change is enough because, after it happens, the window is active and the other events happen serially.

Reply:
Ah, that makes sense. Could you add that explanation to the comments?

Reply:
> Could you add that explanation to the comments?

I already did :) I also added two instances of this comment for the code reader:

```c++
// see preset2Horizontal() for an explanation
```

Reply:
You read these codes more that the rest of us; feel free to add to the branch the explanation you see fit. Sorry, I don't know what you have in mind — to me, that comment explains everything concisely.

Reply:
I added a commit with expanded explanations. Could you check if I understand you correctly?

Reply:
> Could you check if I understand you correctly?

I think we both understand it correctly :)